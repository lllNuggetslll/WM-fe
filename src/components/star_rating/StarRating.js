import React from "react";
import styled from "styled-components";
import { decideStars, starMap } from "./starRatingUtils";

const StarContainer = styled.span`
  margin-left: 10px;
  font-weight: bold;
  opacity: 0.4;
`;

const StarRating = ({ rating }) => {
  const ratingText = rating.toPrecision(2);
  const stars = decideStars(rating);

  return (
    <div>
      {stars.map(rating => {
        const { color, icon } = starMap[rating];

        return <i style={{ color }} className={icon} />;
      })}
      <StarContainer>{ratingText}</StarContainer>
    </div>
  );
};

export default StarRating;

//   For performance, I could've combined the util here and use a prefilled array
//   to achieve just 1 single loop, I chose not to for readability and testability

//   let numStars = rating;
//   {[1, 1, 1, 1, 1].map(rating => {
//     let num;
//
//     if (numStars <= 0) num = 0;
//     if (numStars > 0 && numStars < 0.25) num = 0;
//     if (numStars >= 0.25 && numStars < 1) num = 0.5;
//     if (numStars >= 1) num = 1;
//
//     numStars = numStars - 1;
//
//     const { color, icon } = starMap[num];
//
//     return <i style={{ color }} className={icon} />;
//   })}
