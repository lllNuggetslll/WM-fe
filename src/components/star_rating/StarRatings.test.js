import React from "react";
import { shallow } from "enzyme";
import StarRating from "./StarRating";

describe("StarRatings", () => {
  it("should render all the stars", () => {
    [1, 2, 3, 4, 5].forEach(num => {
      const wrapper = shallow(<StarRating rating={num} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  it("should render half stars", () => {
    const rating = 4.5;

    const wrapper = shallow(<StarRating {...{ rating }} />);
    expect(wrapper).toMatchSnapshot();
  });
});
