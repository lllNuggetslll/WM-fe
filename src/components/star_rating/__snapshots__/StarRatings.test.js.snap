// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarRatings should render all the stars 1`] = `
<div>
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <styled.span>
    1.0
  </styled.span>
</div>
`;

exports[`StarRatings should render all the stars 2`] = `
<div>
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <styled.span>
    2.0
  </styled.span>
</div>
`;

exports[`StarRatings should render all the stars 3`] = `
<div>
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <styled.span>
    3.0
  </styled.span>
</div>
`;

exports[`StarRatings should render all the stars 4`] = `
<div>
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#e8e8e8",
      }
    }
  />
  <styled.span>
    4.0
  </styled.span>
</div>
`;

exports[`StarRatings should render all the stars 5`] = `
<div>
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <styled.span>
    5.0
  </styled.span>
</div>
`;

exports[`StarRatings should render half stars 1`] = `
<div>
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <i
    className="fas fa-star-half-alt"
    key="NaN"
    style={
      Object {
        "color": "#00CDBD",
      }
    }
  />
  <styled.span>
    4.5
  </styled.span>
</div>
`;
